support device scale factor for GraphicsLayerWC

r-otaka opened the pull request:
#### b6d35ab492c201f0cbd3de238a6af2d3c65c156c
<pre>
support device scale factor for GraphicsLayerWC
<a href="https://bugs.webkit.org/show_bug.cgi?id=196339">https://bugs.webkit.org/show_bug.cgi?id=196339</a>

Reviewed by Fujii Hironori.

support device scale factor for GraphicsLayerWC (and WinCairo).
But, NOT using this function now because of following bugs.
so NOW, MiniBrowser sets page zoom to intrinsic device scale factor and
sets &quot;custom device scale factor&quot; to 1.
This simulates the correspondence of device scale factors.

WebInspector always uses intrinsic device scale factor, so if you use
fractional device scale factor, repaint noises may remain.

Remaining bugs
1. Using fractional device scale factor, rendering noises may remain when
   repaint display by JavaScript, scroll with scrollbar dragging or etc.

* Source/WebKit/GPUProcess/graphics/wc/WCScene.cpp:
* Source/WebKit/Shared/NativeWebMouseEvent.h:
* Source/WebKit/Shared/NativeWebWheelEvent.h:
* Source/WebKit/Shared/PlatformPopupMenuData.h:
* Source/WebKit/Shared/PlatformPopupMenuData.serialization.in:
* Source/WebKit/Shared/win/NativeWebMouseEventWin.cpp:
* Source/WebKit/Shared/win/NativeWebWheelEventWin.cpp:
* Source/WebKit/Shared/win/WebEventFactory.cpp:
* Source/WebKit/Shared/win/WebEventFactory.h:
* Source/WebKit/UIProcess/Automation/win/WebAutomationSessionWin.cpp:
* Source/WebKit/UIProcess/Inspector/win/WebInspectorUIProxyWin.cpp:
* Source/WebKit/UIProcess/cairo/BackingStoreCairo.cpp:
* Source/WebKit/UIProcess/wc/DrawingAreaProxyWC.cpp:
* Source/WebKit/UIProcess/wc/DrawingAreaProxyWC.h:
* Source/WebKit/UIProcess/win/PageClientImpl.cpp:
* Source/WebKit/UIProcess/win/WebContextMenuProxyWin.cpp:
* Source/WebKit/UIProcess/win/WebPopupMenuProxyWin.cpp:
* Source/WebKit/UIProcess/win/WebPopupMenuProxyWin.h:
* Source/WebKit/UIProcess/win/WebView.cpp:
* Source/WebKit/UIProcess/win/WebView.h:
* Source/WebKit/WebProcess/WebCoreSupport/win/WebPopupMenuWin.cpp:
* Source/WebKit/WebProcess/WebPage/DrawingArea.h:
* Source/WebKit/WebProcess/WebPage/DrawingArea.messages.in:
* Source/WebKit/WebProcess/WebPage/wc/DrawingAreaWC.cpp:
* Source/WebKit/WebProcess/WebPage/wc/DrawingAreaWC.h:
* Source/WebKit/WebProcess/WebPage/wc/GraphicsLayerWC.cpp:
* Source/WebKit/WebProcess/WebPage/wc/GraphicsLayerWC.h:
* Source/WebKit/WebProcess/WebPage/wc/WCUpdateInfo.h:
* Source/WebKit/WebProcess/WebPage/wc/WCUpdateInfo.serialization.in:
* Tools/MiniBrowser/win/WebKitBrowserWindow.cpp:

Canonical link: <a href="https://commits.webkit.org/279794@main">https://commits.webkit.org/279794@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bb094f1f254448d4a98eb0cbb2ebcab493538ca9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54595 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/34014 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/7167 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/57876 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5329 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/56897 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/41553 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/5331 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/44216 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/60/builds/3585 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/56690 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/32159 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47286 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25342 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/28955 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/4616 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3469 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/50706 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/4832 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59466 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/29828 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/4972 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/51640 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/30978 "Built successfully") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47372 "Exiting early after 10 failures. 10 tests run. 4 failures") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/51017 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/31956 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8075 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/30766 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

khei4 commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -1283,7 +1283,7 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
     void scaleView(double scale);
     void setShouldScaleViewToFitDocument(bool);
-    
+
```
I guess this new line is added unintentionally, we could narrow this patch's scope into WC and Win if we remove this line. ;)

r-otaka replied:
yeah, I will delete it, thank-you.

khei4 commented on Source/WebKit/UIProcess/win/WebView.cpp:
```diff
@@ -544,22 +550,24 @@ LRESULT WebView::onPrintClientEvent(HWND hWnd, UINT, WPARAM wParam, LPARAM, bool
+    if (!m_page)
+        m_viewSize = expandedIntSize(FloatSize(LOWORD(lParam), HIWORD(lParam)) / deviceScaleFactorForWindow(hwnd));
+    else
+        m_viewSize = expandedIntSize(FloatSize(LOWORD(lParam), HIWORD(lParam)) / m_page->deviceScaleFactor());
```
nits: We can reduce redundancy by using ternary op
```suggestion
    float deviceScaleFactor = m_page ? m_page->deviceScaleFactor() : deviceScaleFactorForWindow(hwnd);
    m_viewSize = expandedIntSize(FloatSize(LOWORD(lParam), HIWORD(lParam)) / deviceScaleFactor);
```

r-otaka replied:
My code has fewer variables than your suggested code.
So, I think that PCs can run my code faster than yours.
Does anyone have any ideas about which is good?

khei4 replied:
Of Course this is style nits, but it seems like the code emitted by clang/gcc don't have any difference between ternary version and original one, if we use any optimization level over 1. https://godbolt.org/z/c87YTf3na

r-otaka replied:
I fixed this like yours.
Thank-you.

fujii commented on Source/WebKit/UIProcess/cairo/BackingStoreCairo.cpp:
```diff
@@ -101,8 +101,8 @@ void BackingStore::incorporateUpdate(UpdateInfo&& updateInfo)
 #if ASSERT_ENABLED
-    IntSize updateSize = updateInfo.updateRectBounds.size();
-    updateSize.scale(m_deviceScaleFactor);
+    IntSize updateSize = { static_cast<int>(ceilf(updateInfo.updateRectBounds.size().width() * m_deviceScaleFactor)),
```
I think this can be `IntSize updateSize = expandedIntSize(updateInfo.updateRectBounds.size() * m_deviceScaleFactor);`

r-otaka replied:
That seems smart!
I will change the codes.

fujii commented on Source/WebKit/WebProcess/WebPage/wc/GraphicsLayerWC.cpp:
```diff
@@ -584,6 +591,7 @@ void GraphicsLayerWC::flushCompositingStateForThisLayerOnly()
         update.background.color = backgroundColor();
         if (drawsContent() && contentsAreVisible()) {
             update.background.hasBackingStore = true;
+            update.background.backingStoreSize = WebCore::expandedIntSize(scaled(size(), deviceScaleFactor()));
```
I think `size() * deviceScaleFactor()` is simpler than `scaled(size(), deviceScaleFactor())`  .

r-otaka replied:
Okay, I'll change it.

r-otaka committed:
support device scale factor for GraphicsLayerWC
https://bugs.webkit.org/show_bug.cgi?id=196339

Reviewed by Fujii Hironori.

support device scale factor for GraphicsLayerWC (and WinCairo).
But, NOT using this function now because of following bugs.
so NOW, MiniBrowser sets page zoom to intrinsic device scale factor and
sets "custom device scale factor" to 1.
This simulates the correspondence of device scale factors.

WebInspector always uses intrinsic device scale factor, so if you use
fractional device scale factor, repaint noises may remain.

Remaining bugs
1. Using fractional device scale factor, rendering noises may remain when
   repaint display by JavaScript, scroll with scrollbar dragging or etc.

* Source/WebKit/GPUProcess/graphics/wc/WCScene.cpp:
* Source/WebKit/Shared/NativeWebMouseEvent.h:
* Source/WebKit/Shared/NativeWebWheelEvent.h:
* Source/WebKit/Shared/PlatformPopupMenuData.h:
* Source/WebKit/Shared/PlatformPopupMenuData.serialization.in:
* Source/WebKit/Shared/win/NativeWebMouseEventWin.cpp:
* Source/WebKit/Shared/win/NativeWebWheelEventWin.cpp:
* Source/WebKit/Shared/win/WebEventFactory.cpp:
* Source/WebKit/Shared/win/WebEventFactory.h:
* Source/WebKit/UIProcess/Automation/win/WebAutomationSessionWin.cpp:
* Source/WebKit/UIProcess/Inspector/win/WebInspectorUIProxyWin.cpp:
* Source/WebKit/UIProcess/cairo/BackingStoreCairo.cpp:
* Source/WebKit/UIProcess/wc/DrawingAreaProxyWC.cpp:
* Source/WebKit/UIProcess/wc/DrawingAreaProxyWC.h:
* Source/WebKit/UIProcess/win/PageClientImpl.cpp:
* Source/WebKit/UIProcess/win/WebContextMenuProxyWin.cpp:
* Source/WebKit/UIProcess/win/WebPopupMenuProxyWin.cpp:
* Source/WebKit/UIProcess/win/WebPopupMenuProxyWin.h:
* Source/WebKit/UIProcess/win/WebView.cpp:
* Source/WebKit/UIProcess/win/WebView.h:
* Source/WebKit/WebProcess/WebCoreSupport/win/WebPopupMenuWin.cpp:
* Source/WebKit/WebProcess/WebPage/DrawingArea.h:
* Source/WebKit/WebProcess/WebPage/DrawingArea.messages.in:
* Source/WebKit/WebProcess/WebPage/wc/DrawingAreaWC.cpp:
* Source/WebKit/WebProcess/WebPage/wc/DrawingAreaWC.h:
* Source/WebKit/WebProcess/WebPage/wc/GraphicsLayerWC.cpp:
* Source/WebKit/WebProcess/WebPage/wc/GraphicsLayerWC.h:
* Source/WebKit/WebProcess/WebPage/wc/WCUpdateInfo.h:
* Source/WebKit/WebProcess/WebPage/wc/WCUpdateInfo.serialization.in:
* Tools/MiniBrowser/win/WebKitBrowserWindow.cpp:

Canonical link: https://commits.webkit.org/279794@main
